File: src/AttentionGuardian.Core/ScheduleManagement.cs
namespace AttentionGuardian.Core;

public sealed record MandatoryTodoGroup(
    DateTimeOffset OriginalStart,
    IReadOnlyList<ScheduledTodo> Todos);

public sealed record ScheduleReorderResult(
    IReadOnlyList<ScheduledTodo> ScheduledTodos,
    int ActualIndex,
    bool UsedFallbackPosition,
    bool HasRolloverToNextDay);

public static class ScheduleManagement
{
    public const string BreakTitle = "休息";

    public static IReadOnlyList<MandatoryTodoGroup> FindMandatoryGroups(
        IEnumerable<ScheduledTodo> schedule)
    {
        var ordered = ValidateAndOrder(schedule)
            .Where(todo => todo.IsMandatory)
            .ToArray();
        var groups = new List<MandatoryTodoGroup>();
        var current = new List<ScheduledTodo>();
        DateTimeOffset groupEnd = default;

        foreach (var todo in ordered)
        {
            if (current.Count == 0 || todo.TimeRange.Start <= groupEnd)
            {
                current.Add(todo);
                if (todo.TimeRange.End > groupEnd)
                {
                    groupEnd = todo.TimeRange.End;
                }

                continue;
            }

            AddGroupWhenContinuous(groups, current);
            current = [todo];
            groupEnd = todo.TimeRange.End;
        }

        AddGroupWhenContinuous(groups, current);
        return groups;
    }

    public static ScheduleReorderResult Reorder(
        IEnumerable<ScheduledTodo> schedule,
        Guid todoId,
        int requestedIndex)
    {
        var ordered = ValidateAndOrder(schedule);
        if (requestedIndex < 0 || requestedIndex >= ordered.Count)
        {
            throw new ArgumentOutOfRangeException(nameof(requestedIndex));
        }

        var moving = ordered.SingleOrDefault(todo => todo.Id == todoId)
            ?? throw new ArgumentException("The selected todo does not exist.", nameof(todoId));

        if (moving.IsMandatory)
        {
            return ReorderMandatoryWithinGroup(ordered, moving, requestedIndex);
        }

        var desired = ordered.Where(todo => todo.Id != todoId).ToList();
        desired.Insert(requestedIndex, moving);
        var rebuilt = RebuildFromAnchor(desired, ordered[0].TimeRange.Start);
        var actualIndex = IndexOf(rebuilt, todoId);
        return new(
            rebuilt,
            actualIndex,
            actualIndex != requestedIndex,
            HasRollover(ordered, rebuilt));
    }

    public static IReadOnlyList<ScheduledTodo> Delete(
        IEnumerable<ScheduledTodo> schedule,
        Guid todoId)
    {
        var ordered = ValidateAndOrder(schedule);
        var deletedIndex = ordered.FindIndex(todo => todo.Id == todoId);
        if (deletedIndex < 0)
        {
            throw new ArgumentException("The selected todo does not exist.", nameof(todoId));
        }

        var deleted = ordered[deletedIndex];
        var before = ordered.Take(deletedIndex).ToList();
        var after = ordered.Skip(deletedIndex + 1).ToList();
        var rebuiltAfter = RebuildFromAnchor(after, deleted.TimeRange.Start);
        return before
            .Concat(rebuiltAfter)
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToArray();
    }

    public static IReadOnlyList<ScheduledTodo> Edit(
        IEnumerable<ScheduledTodo> schedule,
        Guid todoId,
        string title,
        TimeSpan duration,
        bool isMandatory)
    {
        var ordered = ValidateAndOrder(schedule);
        var index = ordered.FindIndex(todo => todo.Id == todoId);
        if (index < 0)
        {
            throw new ArgumentException("The selected todo does not exist.", nameof(todoId));
        }

        if (ordered[index].Title == BreakTitle && title != BreakTitle)
        {
            throw new InvalidOperationException("Break events cannot be renamed.");
        }

        var edited = ordered[index].Edit(title, duration, isMandatory);
        var desired = ordered.ToList();
        desired[index] = edited;
        return RebuildFromAnchor(desired, ordered[0].TimeRange.Start);
    }

    public static ScheduleTrialResult InsertBreak(
        IEnumerable<ScheduledTodo> schedule,
        Guid id,
        DateTimeOffset start,
        TimeSpan duration,
        long currentSelectionPriority = 0)
    {
        var breakTodo = new ScheduledTodo(
            id,
            BreakTitle,
            new TimeRange(start, start + duration),
            isMandatory: false,
            currentSelectionPriority);
        return ScheduleTrial.Insert(schedule, breakTodo);
    }

    private static ScheduleReorderResult ReorderMandatoryWithinGroup(
        List<ScheduledTodo> ordered,
        ScheduledTodo moving,
        int requestedIndex)
    {
        var group = FindMandatoryGroups(ordered)
            .SingleOrDefault(candidate => candidate.Todos.Any(todo => todo.Id == moving.Id));
        if (group is null)
        {
            throw new InvalidOperationException(
                "A mandatory todo can only move inside a continuous mandatory group.");
        }

        var groupIndices = group.Todos
            .Select(todo => ordered.FindIndex(candidate => candidate.Id == todo.Id))
            .Order()
            .ToArray();
        if (!groupIndices.Contains(requestedIndex))
        {
            throw new InvalidOperationException(
                "A mandatory todo cannot move outside its continuous group.");
        }

        var reorderedGroup = group.Todos.Where(todo => todo.Id != moving.Id).ToList();
        var targetInGroup = Array.IndexOf(groupIndices, requestedIndex);
        reorderedGroup.Insert(targetInGroup, moving);

        var cursor = group.OriginalStart;
        var replacements = new Dictionary<Guid, ScheduledTodo>();
        foreach (var todo in reorderedGroup)
        {
            var shifted = todo.MoveTo(cursor);
            replacements[todo.Id] = shifted;
            cursor = shifted.TimeRange.End;
        }

        var desired = ordered
            .Select(todo => replacements.GetValueOrDefault(todo.Id, todo))
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToList();
        var rebuilt = RebuildFromAnchor(desired, ordered[0].TimeRange.Start);
        return new(
            rebuilt,
            IndexOf(rebuilt, moving.Id),
            false,
            HasRollover(ordered, rebuilt));
    }

    private static IReadOnlyList<ScheduledTodo> RebuildFromAnchor(
        IReadOnlyList<ScheduledTodo> desiredOrder,
        DateTimeOffset anchor)
    {
        var mandatory = desiredOrder
            .Where(todo => todo.IsMandatory)
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToArray();
        var rebuilt = new List<ScheduledTodo>(desiredOrder.Count);
        var cursor = anchor;

        foreach (var todo in desiredOrder)
        {
            if (todo.IsMandatory)
            {
                rebuilt.Add(todo);
                if (todo.TimeRange.End > cursor)
                {
                    cursor = todo.TimeRange.End;
                }

                continue;
            }

            var shifted = todo.MoveTo(cursor);
            foreach (var blocker in mandatory)
            {
                if (shifted.TimeRange.Overlaps(blocker.TimeRange))
                {
                    shifted = shifted.MoveTo(blocker.TimeRange.End);
                }
            }

            rebuilt.Add(shifted);
            cursor = shifted.TimeRange.End;
        }

        return rebuilt
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToArray();
    }

    private static List<ScheduledTodo> ValidateAndOrder(
        IEnumerable<ScheduledTodo> schedule)
    {
        ArgumentNullException.ThrowIfNull(schedule);
        var items = schedule.ToArray();
        if (items.Any(todo => todo is null))
        {
            throw new ArgumentException("The schedule cannot contain null values.", nameof(schedule));
        }

        if (items.Select(todo => todo.Id).Distinct().Count() != items.Length)
        {
            throw new ArgumentException(
                "The schedule cannot contain duplicate identifiers.",
                nameof(schedule));
        }

        return items
            .OrderBy(todo => todo.TimeRange.Start)
            .ThenBy(todo => todo.TimeRange.End)
            .ThenBy(todo => todo.Id)
            .ToList();
    }

    private static void AddGroupWhenContinuous(
        ICollection<MandatoryTodoGroup> groups,
        IReadOnlyList<ScheduledTodo> current)
    {
        if (current.Count >= 2)
        {
            groups.Add(new(current[0].TimeRange.Start, current.ToArray()));
        }
    }

    private static int IndexOf(IReadOnlyList<ScheduledTodo> todos, Guid id)
    {
        for (var index = 0; index < todos.Count; index++)
        {
            if (todos[index].Id == id)
            {
                return index;
            }
        }

        return -1;
    }

    private static bool HasRollover(
        IReadOnlyList<ScheduledTodo> original,
        IReadOnlyList<ScheduledTodo> rebuilt)
    {
        var originalById = original.ToDictionary(todo => todo.Id);
        return rebuilt.Any(
            todo => todo.ScheduleDate > originalById[todo.Id].ScheduleDate);
    }
}

File: apple/AttentionGuardianDomain/Sources/AttentionGuardianDomain/ScheduleManagement.swift
import Foundation

public enum ScheduleManagementError: Error, Equatable {
    case todoNotFound
    case invalidRequestedIndex
    case mandatoryMoveOutsideContinuousGroup
    case breakCannotBeRenamed
}

public enum StartTimeConflictResolution: String, Sendable {
    case moveExistingAfterEdited
    case truncateExistingAtNewStart
}

public enum StartTimeEditRejection: String, Sendable {
    case none
    case conflictResolutionRequired
    case mandatoryTodoOccupiesNewStart
}

public struct ScheduleReorderResult: Equatable, Sendable {
    public let scheduledTodos: [ScheduledTodo]
    public let actualIndex: Int
    public let usedFallbackPosition: Bool
}

public struct ScheduleStartEditResult: Equatable, Sendable {
    public let scheduledTodos: [ScheduledTodo]
    public let rejection: StartTimeEditRejection
    public let conflictingTodoId: UUID?
}

public enum ScheduleManagement {
    public static let breakTitle = "休息"

    public static func mandatoryContinuousGroups(
        _ schedule: [ScheduledTodo]
    ) -> [[ScheduledTodo]] {
        mandatoryGroups(schedule)
    }

    public static func reorder(
        _ schedule: [ScheduledTodo],
        todoId: UUID,
        requestedIndex: Int
    ) throws -> ScheduleReorderResult {
        let ordered = stable(schedule)
        guard ordered.indices.contains(requestedIndex) else {
            throw ScheduleManagementError.invalidRequestedIndex
        }
        guard let moving = ordered.first(where: { $0.id == todoId }) else {
            throw ScheduleManagementError.todoNotFound
        }

        if moving.isMandatory {
            return try reorderMandatory(
                ordered,
                moving: moving,
                requestedIndex: requestedIndex)
        }

        var desired = ordered.filter { $0.id != todoId }
        desired.insert(moving, at: requestedIndex)
        let rebuilt = try rebuild(desired, anchor: ordered[0].start)
        let actualIndex = rebuilt.firstIndex(where: { $0.id == todoId })!
        return ScheduleReorderResult(
            scheduledTodos: rebuilt,
            actualIndex: actualIndex,
            usedFallbackPosition: actualIndex != requestedIndex)
    }

    public static func delete(
        _ schedule: [ScheduledTodo],
        todoId: UUID
    ) throws -> [ScheduledTodo] {
        let ordered = stable(schedule)
        guard let index = ordered.firstIndex(where: { $0.id == todoId }) else {
            throw ScheduleManagementError.todoNotFound
        }
        let before = Array(ordered[..<index])
        let after = Array(ordered[(index + 1)...])
        let rebuiltAfter = try rebuild(after, anchor: ordered[index].start)
        return stable(before + rebuiltAfter)
    }

    public static func edit(
        _ schedule: [ScheduledTodo],
        todoId: UUID,
        title: String,
        duration: TimeInterval,
        isMandatory: Bool
    ) throws -> [ScheduledTodo] {
        let ordered = stable(schedule)
        guard let index = ordered.firstIndex(where: { $0.id == todoId }) else {
            throw ScheduleManagementError.todoNotFound
        }
        if ordered[index].title == breakTitle && title != breakTitle {
            throw ScheduleManagementError.breakCannotBeRenamed
        }

        let current = ordered[index]
        var desired = ordered
        desired[index] = try ScheduledTodo(
            id: current.id,
            title: title,
            start: current.start,
            end: current.start.addingTimeInterval(duration),
            utcOffsetSeconds: current.utcOffsetSeconds,
            isMandatory: isMandatory,
            currentSelectionPriority: current.currentSelectionPriority)
        return try rebuild(desired, anchor: ordered[0].start)
    }

    public static func insertBreak(
        into schedule: [ScheduledTodo],
        id: UUID,
        start: Date,
        duration: TimeInterval,
        utcOffsetSeconds: Int = 0,
        currentSelectionPriority: Int64 = 0
    ) throws -> ScheduleTrialResult {
        let rest = try ScheduledTodo(
            id: id,
            title: breakTitle,
            start: start,
            end: start.addingTimeInterval(duration),
            utcOffsetSeconds: utcOffsetSeconds,
            isMandatory: false,
            currentSelectionPriority: currentSelectionPriority)
        return try ScheduleTrial.insert(rest, into: schedule)
    }

    public static func editStart(
        _ schedule: [ScheduledTodo],
        todoId: UUID,
        newStart: Date,
        conflictResolution: StartTimeConflictResolution?
    ) throws -> ScheduleStartEditResult {
        let ordered = stable(schedule)
        guard let moving = ordered.first(where: { $0.id == todoId }) else {
            throw ScheduleManagementError.todoNotFound
        }
        let others = ordered.filter { $0.id != todoId }
        let occupying = others.first { $0.contains(newStart) }

        if let occupying, occupying.isMandatory {
            return ScheduleStartEditResult(
                scheduledTodos: ordered,
                rejection: .mandatoryTodoOccupiesNewStart,
                conflictingTodoId: occupying.id)
        }
        if let occupying, conflictResolution == nil {
            return ScheduleStartEditResult(
                scheduledTodos: ordered,
                rejection: .conflictResolutionRequired,
                conflictingTodoId: occupying.id)
        }

        var prefix = others.filter {
            $0.id != occupying?.id && $0.end <= newStart
        }
        let prefixIds = Set(prefix.map(\.id))
        let remaining = others.filter {
            $0.id != occupying?.id && !prefixIds.contains($0.id)
        }
        let moved = try moving.moved(to: newStart)

        if let occupying, conflictResolution == .truncateExistingAtNewStart {
            prefix.append(try ScheduledTodo(
                id: occupying.id,
                title: occupying.title,
                start: occupying.start,
                end: newStart,
                utcOffsetSeconds: occupying.utcOffsetSeconds,
                isMandatory: occupying.isMandatory,
                currentSelectionPriority: occupying.currentSelectionPriority))
        }

        var suffix = [moved]
        if let occupying, conflictResolution == .moveExistingAfterEdited {
            suffix.append(occupying)
        }
        suffix.append(contentsOf: remaining)
        let rebuilt = try rebuild(suffix, anchor: newStart)
        return ScheduleStartEditResult(
            scheduledTodos: stable(prefix + rebuilt),
            rejection: .none,
            conflictingTodoId: occupying?.id)
    }

    private static func reorderMandatory(
        _ ordered: [ScheduledTodo],
        moving: ScheduledTodo,
        requestedIndex: Int
    ) throws -> ScheduleReorderResult {
        guard let group = mandatoryGroups(ordered).first(
            where: { $0.contains(where: { $0.id == moving.id }) })
        else {
            throw ScheduleManagementError.mandatoryMoveOutsideContinuousGroup
        }
        let indices = group.compactMap { grouped in
            ordered.firstIndex(where: { $0.id == grouped.id })
        }.sorted()
        guard indices.contains(requestedIndex) else {
            throw ScheduleManagementError.mandatoryMoveOutsideContinuousGroup
        }

        var reordered = group.filter { $0.id != moving.id }
        reordered.insert(moving, at: indices.firstIndex(of: requestedIndex)!)
        var cursor = group.map(\.start).min()!
        var replacements: [UUID: ScheduledTodo] = [:]
        for todo in reordered {
            let shifted = try todo.moved(to: cursor)
            replacements[todo.id] = shifted
            cursor = shifted.end
        }
        let desired = stable(ordered.map { replacements[$0.id] ?? $0 })
        let rebuilt = try rebuild(desired, anchor: ordered[0].start)
        return ScheduleReorderResult(
            scheduledTodos: rebuilt,
            actualIndex: rebuilt.firstIndex(where: { $0.id == moving.id })!,
            usedFallbackPosition: false)
    }

    private static func mandatoryGroups(
        _ schedule: [ScheduledTodo]
    ) -> [[ScheduledTodo]] {
        let mandatory = stable(schedule.filter(\.isMandatory))
        var groups: [[ScheduledTodo]] = []
        var current: [ScheduledTodo] = []
        var groupEnd: Date?
        for todo in mandatory {
            if current.isEmpty || todo.start <= groupEnd! {
                current.append(todo)
                groupEnd = max(groupEnd ?? todo.end, todo.end)
            } else {
                if current.count >= 2 { groups.append(current) }
                current = [todo]
                groupEnd = todo.end
            }
        }
        if current.count >= 2 { groups.append(current) }
        return groups
    }

    private static func rebuild(
        _ desired: [ScheduledTodo],
        anchor: Date
    ) throws -> [ScheduledTodo] {
        let mandatory = stable(desired.filter(\.isMandatory))
        var result: [ScheduledTodo] = []
        var cursor = anchor
        for todo in desired {
            if todo.isMandatory {
                result.append(todo)
                cursor = max(cursor, todo.end)
                continue
            }
            var shifted = try todo.moved(to: cursor)
            for blocker in mandatory where overlaps(shifted, blocker) {
                shifted = try shifted.moved(to: blocker.end)
            }
            result.append(shifted)
            cursor = shifted.end
        }
        return stable(result)
    }

    private static func overlaps(
        _ left: ScheduledTodo,
        _ right: ScheduledTodo
    ) -> Bool {
        left.start < right.end && right.start < left.end
    }

    private static func stable(_ schedule: [ScheduledTodo]) -> [ScheduledTodo] {
        schedule.sorted {
            if $0.start != $1.start { return $0.start < $1.start }
            if $0.end != $1.end { return $0.end < $1.end }
            return $0.id.uuidString.lowercased() < $1.id.uuidString.lowercased()
        }
    }
}
